### DISCUSSION TITLE
How to force xpra to render window without client connection?

### DISCUSSION THREAD

**Discussion by Gavin1937** [Xpra-org/xpra#3963] [Q&A] [answered]
I'm using xpra as the way to run normal browser (with GUI) inside a headless server for automation.

However, the browser that xpra opens up always show a white line like following:
(taken using selenium)

![1691975945](https://github.com/Xpra-org/xpra/assets/71205842/9e36dcdc-4fa1-47f6-ac7c-6b7714ed5c7e)

I guess this is because xpra server will not render browser window when there are no client connection.

I can verify that because I can receive the complete webpage html while no client is connecting to the server, and my script cannot click on any elements on the page because xpra is being lazy. But once I connect to the server, everything works fine.

Is there a way to force xpra to render window without client connection?

I'm running xpra server on a Ubuntu 22.04 machine.

**Comment by totaam** [comment, marked as answer]
> and my script cannot click on any elements on the page because xpra is being lazy

No idea what that means.

My guess is that the browser is not mapped on the virtual screen, hence why you can't interact with it.
Try force mapping it with `xdotool`.

**Reply by Gavin1937** [reply]
I fixed it with `xdotool`

my origial script to launch browser is:

```sh
xpra start :500 --start="google-chrome --remote-debugging-port=9000 --user-data-dir='/path/to/datadir'"
```

Then, when I take a screenshot with selenium on this browser setup, I get a 500x1 px white bar showed above.

Now, I use `xdotool` to force mapping xpra, and my browser launch script become:

```sh
XPRA_PORT="500"

# start xpra desktop & wait for it a bit
xpra start-desktop ":$XPRA_PORT" --start="google-chrome --remote-debugging-port=9000 --user-data-dir='/path/to/datadir'" &&
sleep 10 &&

# get xpra google-chrome pid
PID=$(ps aux | grep -e xpra.*google\-chrome | grep -v grep | awk '{print $2}')

# setup env variables for xdotool
export DISPLAY=":$XPRA_PORT"
export XAUTHORITY=$(cat "/run/user/1000/xpra/$XPRA_PORT/xauthority")

# use xdotool to map xpra desktop on the virtual screen, maximize window, & focus on the window
xdotool search --pid $PID --class google-chrome windowmap --sync
xdotool search --pid $PID --class google-chrome windowsize --sync 100% 100%
xdotool search --pid $PID --class google-chrome windowfocus

echo -n "getwindowname = "
xdotool search --pid $PID --class google-chrome getwindowname
```

**Comment by totaam** [comment]
> `XPRA_PORT="500"`

FYI: This is not a port, it is a display number.

> `sleep 10 &&`

Using sleep is unreliable.
You could wait until `xpra info` shows that a window exists.

---

IMO: A better approach would be to put everything in a single script you run with `--start`.
This would have access to `$DISPLAY` no matter what value is actually used. Saving you having to choose one that may or may not be available.

> `PID=$(ps aux | grep -e xpra.*google\-chrome | grep -v grep | awk '{print $2}')`

You should use probably `pidof`.

**Reply by Gavin1937** [reply]
thanks for the correction & tips, and also thanks for creating this great software!
